### ISSUE TITLE
Magit Log buffer: pick select commits for diff ?

### ISSUE THREAD

**Issue by alfredking1988** [magit/magit#3441]
My apologies if this issue has been raised before.

In `Magit Log` buffer, I want to just pick some commits and see the diff. I know if these commits are consecutive, I can use `set-mark-command` to select them.

But when they are not consecutive, the mark doesn't work.  So I try to use the `magit-log-select-pick` just like what I do in the rebase. But it also doesn't work.

Is it possible to have the pick feature in the `Magit-Log ` buffer ?

Thanks ~

**Comment by tarsius** [comment]
> But when they are not consecutive, the mark doesn't work.

Did you actually try? This is supposed to work. It doesn't matter if there are unrelated commits in between, all that matters are the commits at the end of the region.

**Comment by alfredking1988** [comment]
Sorry for my poor English.

What I mean is if we have some commits:

`commit 5`
`commit 4`
`commit 3`
`commit 2`
`commit 1`

And I want to view the diff from `commit 1` all the way to `commit 5`, then I can just use the `set-mark-command` to mark the whole region and then type `d r`

Now suppose that `commit 1` , `commit 3` and `commit 5` are made by me, and I want to view them, how can I only select these three commits without select the `commit 2` and `commit 4` in between ?

**Comment by alfredking1988** [comment]
After some search,  I find a way to do this.

Continuing the above example,  first create a `tmp` branch, then interactive `rebase`. In the `Rebase buffer`, type `k` to drop the `commit 2` and `commit 4`.  Type `C-c C-c` to finish.

After that, we have a branch that only contains `commit 1`, `commit 3` and `commit 5`. Now we can mark them to view the diff.

When we finish view the diff, we can delete this tmp branch.

This works, but it's not convenient. Is it possible for Magit to have a new feature that automatically do this for us ?  (Just like the PyCharm IDE, it can just select the commits we want to view and see the diff.)

**Comment by tarsius** [comment]
Can you point me to the place in the PyCharm IDE documentation where that feature is described? Some screenshots might also help.

I have some doubts that it really does what you think it does. At least it would only sometimes be able to do it. In your example imagine `commit 2` adds three completely new lines and `commit 3` modifies the second of these lines. If you tell some tool to show you the changes introduced by `commit 1 + 3 + 5` *in a single diff*, then it could do that using rebase as you have done manually, but in this example `commit 3` is meaningless without `commit 2` and the tool would have to have you resolve the issue or just ignore `commit 3` and hope that won't make other subsequent commits meaningless too.

**Comment by tarsius** [comment]
I suspect what PyCharm actually does is something like `git log --oneline -u --author=YOU`, which you can do in Magit using `l l` - you will have to unset `--graph` and set `--patch` and `--author` in the popup and you might also want to reduce the value of `-n`.

**Comment by alfredking1988** [comment]
Sorry, I do not use PyCharm, I saw its feature from others. I will install the PyCharm to give a try, to see what really happen.

Indeed, the `commit 3` is meaningless without `commit 2 ` in the scenario you described above.

I think your are right, the PyCharm just show the something like `git log --author=...`.
But the PyCharm does show them as a `single` diff, in a side-by-side fashion.

The side-by-side fashion can be easily achieved by `magit-ediff-dwim`.

But the Magit `l l` log with `--author` just show the separate diff, not a _single_ diff

Is there any way to show the commits in a `single` diff ?

**Comment by tarsius** [comment]
> Is there any way to show the commits in a _single_ diff ?

The only way to achieve that is for there to *actually* be a single diff, which isn't the case until after rewriting history. Because that is very likely to fail and require human intervention I do not want to provide a command that attempts this in the background.

> The side-by-side fashion can be easily achieved by `magit-ediff-dwim`.

Ediff can show to versions of a file side-by-side in two separate buffers. It cannot show more than two files (without opening two new buffers per file) or show only the parts of the file that changes.

In other words in order to provide such a side-by-side "single diff" or show side-by-side diffs in logs, Magit cannot use Ediff and would have to grow its own side-by-side diff functionality, which would be a lot of work.

Maybe I will implement that one day, but that is so far off, I won't even keep this issue open. Sorry.

**Comment by alfredking1988** [comment]
Ok, thank you ~

**Comment by alfredking1988** [comment]
Hello, I install WebStorm, it built by the same company who build the PyCharm.

First, I select some commits, the commits are not consecutive.

![pick](https://user-images.githubusercontent.com/34473300/39907275-3235247a-551a-11e8-8831-c7efc8675afa.png)

Then I can view them side-by-side, as a single diff.

![side-by-side](https://user-images.githubusercontent.com/34473300/39907280-38c27126-551a-11e8-9097-e9fa4bd2ebb9.png)

**Comment by tarsius** [comment]
That does *not* show the diff for more than one commit, it ignored the fact that you had more than one commit elected. Since I don't think it can do what you requested it to do, it should have shown an error message instead of just doing something else.

In the second screenshot, in the small window/frame in the middle, the left buffer contains the file as it was in `dee7271` ("magit--insert-modules-overview: fail back to HEAD commit SHA-1") and the right buffer contains the version from `dee7271`, which isn't even one of the commits that you selected.

`dee7271` is the parent of `b662d72`. So what this does is show you the change made by `b662d72` and `b662d72` only. Also note that in the top right corner it says "1 difference".

**Comment by alfredking1988** [comment]
> In the second screenshot, in the small window/frame in the middle, the left buffer contains the file as it was in dee7271 , ... and the right buffer contains the version from dee7271

Yes, the left buffer is from `dee7271`. But I think the right buffer contains the version from `b662d72`

> Also note that in the top right corner it says "1 difference",  if this really showed the changes made by multiple commits, then you would usually expect more than one difference.

I select 4 commits, only `b662d72` change this file `magit-pop.el`, so it should show the change made by this commit, other commits just change the `magit-popup.org` and `magit-popup.texi`

I can press `F7` to let the IDE show the next file, this is what it show for the `magit-popup.org`:

![org](https://user-images.githubusercontent.com/34473300/39936030-b70e06f4-557d-11e8-9090-878aea94d4e2.png)

Line 11 in the right buffer is

> #+BIND: ox-texinfo...

This introduced in the _lowest_ commit `make: use ox-texinfo+-update-version-strings` I selected, so I think this show the changes made by multiple commits
